internal/mempool/mempool.go

```diff
@@ -872,19 +876,154 @@ func (g *txDescGraph) forEachDescendant(txHash *chainhash.Hash,
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor transaction does not have stats, then the
+			// provided one should not either.
+			ancestorTxHash := *ancestorTxDesc.Tx.Hash()
+			ancestorStats, exists := mv.ancestorStats[ancestorTxHash]
+			if !exists {
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor plus its respective ancestors would exceed the
+			// limit for this transaction, then do not track stats.
+			if ancestorStats.NumAncestors+1 > ancestorTrackingLimit {
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor has more children than the tracking limit,
+			// then stats should not be tracked for the provided transaction.
+			ancestorChildren := mv.txGraph.childrenOf[ancestorTxHash]
+			if len(ancestorChildren) >= ancestorTrackingLimit {
+				canTrackAncestors = false
+				return false
+			}
+
+			aggregateAncestorStats(stats, ancestorTxDesc)
+			return true
+		})
+
+	if !canTrackAncestors {
+		delete(mv.ancestorStats, *txHash)
+		return false
+	} else {
+		mv.ancestorStats[*txHash] = stats
+		return true
+	}
+}
+
 // updateStatsDescendantsRemoved removes the provided transaction's stats
 // from all descendant transactions.
 func (mv *txMiningView) updateStatsDescendantsRemoved(txDesc *mining.TxDesc) {
 	txHash := txDesc.Tx.Hash()
-	txSize := int64(txDesc.Tx.MsgTx().SerializeSize())
-	seen := make(map[chainhash.Hash]struct{})
-	mv.txGraph.forEachDescendant(txHash, seen, func(descendant *mining.TxDesc) {
-		descendantStats := mv.ancestorStats[*descendant.Tx.Hash()]
-		descendantStats.Fees -= txDesc.Fee
-		descendantStats.SizeBytes -= txSize
-		descendantStats.TotalSigOps -= txDesc.TotalSigOps
-		descendantStats.NumAncestors--
-	})
+	if _, hasStats := mv.ancestorStats[*txHash]; !hasStats {
+		// If the transaction does not have ancestor tracking enabled, then
+		// none of its descendants should either.
+		return
+	}
+
+	seenDescendants := make(map[chainhash.Hash]struct{}, ancestorTrackingLimit)
+	mv.txGraph.forEachDescendantPreOrder(txHash, seenDescendants,
+		func(descendant *mining.TxDesc) bool {
+			descendantStats, hasStats := mv.ancestorStats[*descendant.Tx.Hash()]
+			if hasStats && len(seenDescendants) > ancestorTrackingLimit {
+				// If the descendant has stats but we've walked too many,
+				// then stop tracking stats for this descendant and its
+				// descendants.
+				delete(mv.ancestorStats, *txHash)
+				return true
+			}
+
+			// Start tracking the descendant's ancestor stats if it was not
+			// tracked previously. This scenario applies to a transaction
+			// that once had too many ancestors, but is now eligible to
+			// have stats tracked. Note that this call retrieves ancestors
+			// while traversing descendants of another transaction, and that
+			// the time complexity is bound by the maximum number of
+			// ancestors allowed to be tracked in the graph.
+			if !hasStats && mv.maybeUpdateAncestorStats(descendant) {
```

Comment:
This allows for testing too many descendants for eligibility for inclusion in the ancestor-tracked set of transaction.

```suggestion
			if !hasStats && len(seenDescendants) <= ancestorTrackingLimit && mv.maybeUpdateAncestorStats(descendant) {
```